File: 0328-odd-even-linked-list/0328-odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == nullptr) {
            return head;
        }
        ListNode* odd = head;
        if (odd->next == nullptr || odd->next->next == nullptr) {
            return head;
        }
        ListNode* even = head->next;
        while (head && head->next) {
            ListNode* nxt = head->next;
            head->next = nxt->next;
            head = nxt;
        }
        ListNode* p = odd;
        while (p->next) {
            p = p->next;
        }
        p->next = even;
        return odd;
    }
};

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null || head.next.next == null) {
            return head;
        }
        ListNode odd = head;
        ListNode even = head.next;
        while (head != null && head.next != null) {
            ListNode next = head.next;
            head.next = next.next;
            head = next;
        }
        ListNode ptr = odd;
        while (ptr.next != null) {
            ptr = ptr.next;
        }
        ptr.next = even;
        return odd;
    }
}